### PR TITLE
Add MiMa to keep backward binary compatibility under control.

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @gzm0 
I did test manually that it does report errors if I introduce backward binary incompatible changes. This tool is really cool!

File: project/Build.scala [hunk 1]
```diff
@@ -210,7 +215,12 @@ object Build extends sbt.Build {
   lazy val irProject: Project = Project(
       id = "ir",
       base = file("ir"),
-      settings = commonIrProjectSettings
+      settings = commonIrProjectSettings ++ (
+          mimaDefaultSettings
```
Comment (gzm0) [inline project/Build.scala]: commonIrProjectSettings?

Comment (sjrd) [inline project/Build.scala]: I want to keep it close to `previousArtifact`.

Comment (gzm0) [inline project/Build.scala]: Ok...

**Comment by gzm0** [comment]
I am assuming there is a reason `%%` and `%%%!` do not work. (I don't even understand why `%%%` wouldn't work)....

**Comment by sjrd** [comment]
`%%%` would not work because the implementation of the macro is in the same project (i.e., the build project).

`%%` and `%%%!` don't work because, well, they don't work. It appears the MiMa sbt plugin does not honor the cross version of the artifact in `previousArtifact`.

File: project/Build.scala (hunk 1 above)
Comment (sjrd) [inline project/Build.scala]: Well so actually I discovered I can simply put `mimaDefaultSettings` in our `commonSettings`, shared by all projects. Then only `previousArtifact` needs to be set in projects for which we want to check binary compat. When it's not set, it defaults to `None`, and binary compat is ignored.

**Comment by sjrd** [comment]
Hence updated.

**Comment by gzm0** [comment]
LGTM

**Pull request merged**